### PR TITLE
SHA256 experiment

### PR DESCRIPTION
**Pull request by daniel-scs**
This is just to float the idea of pulling in the `sha2` crate, which appears to have been the consensus plan for our Thread needs (for deriving keys) over at issue #579.

I was surprised at the object-size delta:

master:
```
   text    data     bss     dec     hex filename
 133768    4840   27920  166528   28a80 target/thumbv7em-none-eabi/release/imix
```

this experiment:
```
   text    data     bss     dec     hex filename
 141048    4840   27920  173808   2a6f0 target/thumbv7em-none-eabi/release/imix
```

That is, just 7280 additional bytes of text.

The transitive closure of new dependencies is:
```
block-buffer, byte-tools, digest, fake-simd, generic-array, nodrop, sha2, typenum
```

Thoughts?

### CODE REVIEW

**Comment by dadrian** [comment]
I just want to note that #579 discussed HMAC-SHA256, whereas this PR only brings in SHA256. You'll still need to bring in the [HMAC crate](https://crates.io/crates/hmac) to be able to use the hash as part of a hash-based message-authentication code (HMAC), rather than just as a function from "lots of data" to "256-bits of data".

**Comment by alevy** [comment]
We're definitely going to need a Safe-Haskell-like mechanism if we're gonna start important all sorts of crates...

**Comment by alevy** [comment]
Maybe we can hack this together with an additional pass on the `capsules` crate for now...

**Comment by daniel-scs** [comment]
Yeah, has there been any movement to do a Safe-Haskell-like mechanism in the Rust community?  Should we consider doing it ourselves?

And what do you mean by "an additional pass"?

**Comment by daniel-scs** [comment]
@dadrian : Yeah, HMAC is pretty straightforward once we have a hash implementation, so I figure we will write that ourselves if necessary.  But I'll look at the HMAC crate as part of this experiment ...

**Comment by daniel-scs** [comment]
I've added the `hmac` crate and a little test of HMAC-SHA256.

This adds another 2K of text (so a total 10K for sha2+hmac):

```
   text    data     bss     dec     hex filename
 143368    4840   27920  176128   2b000 target/thumbv7em-none-eabi/release/imix
```
And it adds these dependencies:

```
constant_time_eq, crypto-mac, hmac
```

**Comment by alevy** [comment]
Formatting is failing, but I think this isn't meant to merge anyway, right? If so, can we put a comment in the description not to merge it?

**Comment by alevy** [comment]
> has there been any movement to do a Safe-Haskell-like mechanism in the Rust community

There is a way for a crate to disallow itself from using `unsafe` (which we now use in `capsules`) and there is a way to prevent _all_ crates cargo will compile from using `unsafe` (by setting the `RUSTC_FLAGS` environment variable). But no current way to enforce that some dependencies can while others cannot use `unsafe`.

I posted a "pre-RFC" a few months ago for Cargo https://internals.rust-lang.org/t/pre-rfc-cargo-safety-rails/5535 which was received with mixed reviews and haven't had cycles to follow up on it.

> And what do you mean by "an additional pass"?

Basically run rustc manually some how _just_ to verify that only a trusted set of crates use `unsafe` then normally run `xargo` to compile the actual binary.

**Comment by bradjc** [comment]
As a note, of the 11 crates Daniel listed, these repos return results when searching for `unsafe` (note utils repo contains several crates):

- https://github.com/paholg/typenum/search?utf8=%E2%9C%93&q=unsafe&type=
- https://github.com/bluss/arrayvec/search?utf8=%E2%9C%93&q=unsafe&type=
- https://github.com/RustCrypto/utils/search?utf8=%E2%9C%93&q=unsafe&type=
- https://github.com/RustCrypto/hashes/search?utf8=%E2%9C%93&q=unsafe&type=
- https://github.com/fizyk20/generic-array/search?utf8=%E2%9C%93&q=unsafe&type=

So if we include this as a dependency of `capsules` we can't really say that capsules is `unsafe` free. Now I didn't actually look into why they have unsafe in the source, it's possible things aren't as bad as they look. This is an interesting realization though, that so many libraries are subverting the rust typesystem and there is little recourse for developers who probably don't want to audit all of their dependencies.

**Comment by daniel-scs** [comment]
Oh, thanks for checking that out, Brad.

If we go this route, I think we'll just have to make these crates dependencies of the kernel crate (instead of capsules, as I've got it here).

Probably it would be possible to hack this code to avoid the use of unsafe, but in general it seems sad to have to "fork" community libraries whenever we use them.

I'm going to close this PR now as I've got the feedback I need.
